File: purecoincidence/src/main/scala/purecoincidence/Purity.scala
package purecoincidence

object Purity extends App {
  def pure(a: Int, b: Int): Int =
    2*a*a + 3*b    // 2a^2 + 3b

  val p = pure(2, 3)   // => 17

  println((
    p + p,
    pure(2, 3) + p,
    pure(2, 3) + pure(2, 3),
  ))    // => (34,34,34)
}

File: hopefullypure/app/src/main/java/hopefullypure/Purity.java
package hopefullypure;

import static java.lang.System.out;
import static java.lang.String.format;

public class Purity {
    static int pure(int a, int b) {
        return 2*a*a + 3*b;
    }

    public static void main(String[] args) {
        int p = pure(2, 3);
        out.println(
          format("%d, %d, %d",
                 p + p,
                 pure(2, 3) + p,
                 pure(2, 3) + pure(2, 3))); // => 34, 34, 34
    }
}
